source/server/listener_impl.h

```diff
@@ -311,11 +379,29 @@ class ListenerImpl final : public Network::ListenerConfig,
   const std::vector<AccessLog::InstanceSharedPtr>& accessLogs() const override {
     return access_logs_;
   }
+  Event::Dispatcher& dispatcher() override;
+  void rebuildFilterChain(const envoy::config::listener::v3::FilterChain* const& filter_chain,
+                          const std::string& worker_name) override;
+
   uint32_t tcpBacklogSize() const override { return tcp_backlog_size_; }
   Init::Manager& initManager();
   envoy::config::core::v3::TrafficDirection direction() const override {
     return config().traffic_direction();
   }
+  bool
+  containFilterChain(const envoy::config::listener::v3::FilterChain& filter_chain) const override {
+    return filter_chain_manager_.containFilterChain(filter_chain);
+  }
+
+  // Stop rebuilding filter chain on timeout of rebuilder, make the filter chain back to a
+  // placeholder. Will be called by the rebuilder.
+  void
+  stopRebuildingFilterChain(const envoy::config::listener::v3::FilterChain* const& filter_chain);
+
+  // Stop all unfinished rebuilding, will be called when listener update. Because it is possible
+  // that old listener is drained and removed before the filter chain has finished rebuilding. We
+  // should remove the rebuilt filter chains that are temporarily stored inside the placeholder.
+  void stopUnfinishedFilterChainRebuilding();
```

Comment:
rename to cancelAllFilterChainRebuild()?

Reply:
Updated